File: distros/M/Mail-SpamCannibal/DNSBLserver/misc.c
/* misc.c
 *
 * Portions of this file are adapted from LaBrea by Tom Liston <tliston@premmag.com>, Copyright (C) 2001, 2002
 #
 * Copyright 2003 - 2009, Michael Robinton <michael@bizsystems.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include <unistd.h>
#include <time.h>
#include <stdio.h>
#include <signal.h>
#include <syslog.h>
#include <signal.h>
#include <sys/wait.h>
#include "defines.h"
#include "util_pid_func.h"

pid_t zonechild = 0;

void
LogPrint(char *output) {
  extern int oflag;
  time_t current;
  char tbuf[17];

  if(oflag) {
    current = time(NULL);
    strncpy(tbuf,ctime(&current)+4,16);
    strcpy(tbuf+16,"\0");
    printf("%s%s\n",tbuf,output);
  }
  else
    syslog(INFOTYPE, output);
}

void
CleanExit(int sig) {
  extern DBTPD dbtp;
  extern int logopen, fdUDP, fdTCPlisten, fdTCP, parent, unlinkOK;
  extern char str2[], str3[];
  
  if(sig == SIGHUP) {
    LogPrint(str2);
    return;
  }

  /* close the database and environment	*/
  dbtp_close(&dbtp);

  /* close open sockets	*/
  if (fdUDP)
    close(fdUDP);
  if (fdTCPlisten)
    close(fdTCPlisten);
  if (fdTCP)
    close(fdTCP);

  if (unlinkOK)
	unlink(pidpath());	/* remove pid file	*/

  /* shut down syslog connection */
  if(parent)
    LogPrint(str3);
  if(logopen)
    closelog();
  /* error exit if not SIGTERM  */
  if (sig != SIGTERM)
    exit(1);
  exit(0);
}

void
ToggleLogging(int sig) {
  extern int datalog, savedatalog, oflag;
  extern char format5[], str19[], str20[], str21[];
  char logbuffer[255];
  if((datalog == 0) && (savedatalog == 0)) {
    datalog = 1;
    sprintf(logbuffer, str19, sig);
    goto logit;
  } else {
    if(savedatalog != 0) {
      datalog = savedatalog;
      savedatalog = 0;
      sprintf(logbuffer, str20, sig, datalog);
      goto logit;
    } else {
      savedatalog = datalog;
      datalog = 0;
      sprintf(logbuffer, str21, sig, datalog);
      goto logit;
    }
  }
  return;
  logit:
  if(oflag)
    printf(format5, logbuffer);
  else
    syslog(INFOTYPE, logbuffer);
}

void
ZoneDump(int sig)
{
  extern int zone_request;

  zone_request = 1;
  LogPrint(str1); 
}

void
SigHup(int sig)
{
  extern int datalog;
  if (datalog > 1)
    LogPrint(str4);
  return;	/* ignore */
}

void
SigChild(int sig)
{
  extern pid_t zonechild;

  pid_t rv;
  if (sig == SIGCHLD) {
    do {
      rv = waitpid(-1, NULL, WNOHANG);
      if (zonechild && zonechild == rv)
      	zonechild = 0;
    } while (rv > 0);
  }
  return;
}

void
q_handler (int sig)
{
  switch (sig) {
	case SIGALRM	: CleanExit(sig);
			  break;
	case SIGUSR1	: ToggleLogging(sig);
			  break;
	case SIGUSR2	: ZoneDump(sig);
			  break;
	case SIGCHLD	: SigChild(sig);
			  break;
	case SIGHUP	: SigHup(sig);
			  break;
	case SIGINT	:
	case SIGQUIT	:
	case SIGTERM	: CleanExit(sig);
			  break;
  }
}

void
set_signals (void)
{
  extern struct sigaction sa;

  sa.sa_handler = q_handler;
  sigemptyset (&sa.sa_mask);
  sigaddset (&sa.sa_mask, SIGTERM);
  sigaddset (&sa.sa_mask, SIGHUP);
  sigaddset (&sa.sa_mask, SIGUSR1);
  sigaddset (&sa.sa_mask, SIGUSR2);
  sigaddset (&sa.sa_mask, SIGQUIT);
  sigaddset (&sa.sa_mask, SIGINT);
  sigaddset (&sa.sa_mask, SIGALRM);
  sigaddset (&sa.sa_mask, SIGCHLD);
  sa.sa_flags = 0;
  sigaction (SIGCHLD, &sa, NULL);
  sigaction (SIGALRM, &sa, NULL);
  sigaction (SIGINT, &sa, NULL);
  sigaction (SIGQUIT, &sa, NULL);
  sigaction (SIGUSR1, &sa, NULL);
  sigaction (SIGUSR2, &sa, NULL);
  sigaction (SIGTERM, &sa, NULL);
  sigaction (SIGHUP, &sa, NULL);
}

File: distros/G/Glib-Ex-ConnectProperties/devel/misc.pl
#!/usr/bin/perl -w

# Copyright 2007, 2008, 2009, 2010 Kevin Ryde

# This file is part of Glib-Ex-ConnectProperties.
#
# Glib-Ex-ConnectProperties is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 3, or (at your option) any
# later version.
#
# Glib-Ex-ConnectProperties is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Glib-Ex-ConnectProperties.  If not, see <http://www.gnu.org/licenses/>.


use 5.008;
use strict;
use warnings;
use Glib;
# use Gtk2 '-init';
use Scalar::Util;

{
  require Gtk2;
  Gtk2->init;
  my $label = Gtk2::Label->new;
  $label->set_state(0);
  exit 0;
}

{
  my $x = Glib::ParamFlags->new ('readable');
  my $y = Glib::ParamFlags->new ('writable');
  print $x->all ($y, 0);
  exit 0;
}

{
  foreach my $pspec (Gtk2::Gdk::Window->list_properties) {
    print $pspec->get_name,"\n";
  }
  exit 0;
}

{
  Gtk2::HBox->new;
  foreach my $type ('Glib::Float',
                    'Gtk2::HBox',
                    'Gtk2::Gdk::EventMask',
                    'Gtk2::Gdk::CursorType',
                    'Glib::Flags',
                    'Glib::Enum',
                    'Gtk2::PackType') {
    print "$type ancestors: ";
    eval { print join (' ', Glib::Type->list_ancestors ($type)),"\n"; }
      or print "error $@";
  }
  exit 0;
}
{
  my $type = 'Gtk2::Gdk::Rectangle';
  print $type->can('equal') ? "yes\n" : "no\n";
  my $rect = Gtk2::Gdk::Rectangle->new (1,2,3,4);
  print "$rect\n";
  exit 0;
}
{
  my $y = 123;
  my $x = \$y;
  print $x->isa('foo') ? "yes\n" : "no\n";
  print $x->isa('foo') ? "yes\n" : "no\n";
  exit 0;
}
{
  my $type = 'Glib::String';
#   $type = 'Gtk2::Gdk::Color';
#   $type = 'Gtk2::Gdk::EventMask';
#   $type = 'Foo';
  print $type->isa($type) ? "yes\n" : "no\n";
  exit 0;
}
{
  my $type = 'Gtk2::Gdk::Color';
  $type = 'Gtk2::Gdk::EventMask';
  print $type->can('equal') ? "yes\n" : "no\n";
}
{
  my $type = 'Gtk2::Widget';
  my $supertype = 'Gtk2::Object';
  print $type->isa($supertype) ? "yes\n" : "no\n";
}

my $label = Gtk2::Label->new ('Hello');
my $button = Gtk2::CheckButton->new_with_label ('Press');

require Glib::Ex::ConnectProperties;
my $conn = Glib::Ex::ConnectProperties->new ([$label,'sensitive'],
                                             [$button,'active']);

print "\nfrozen:\n";
$label->freeze_notify;
$button->set (active => 1);
$button->set (active => 0);

print "\nthaw:\n";
$label->thaw_notify;

exit 0;
